### PR TITLE
Tests/Kernel: Properly synchronize threads in TestTCPSocket

### PR DESCRIPTION
**Pull request by DanShaders**
We should wait for a server thread to actually listen(2) a server socket before trying to connect to it. This hopefully should make the test less flaky.

### CODE REVIEW

File: Tests/Kernel/TestTCPSocket.cpp [hunk 1]
```diff
@@ -49,7 +52,11 @@ static void* server_handler(void*)
 static pthread_t start_tcp_server()
 {
     pthread_t thread;
-    int rc = pthread_create(&thread, nullptr, server_handler, nullptr);
+    sem_t accept_semaphore;
+    sem_init(&accept_semaphore, 0, 0);
+    int rc = pthread_create(&thread, nullptr, server_handler, &accept_semaphore);
+    VERIFY(sem_wait(&accept_semaphore) == 0);
+    VERIFY(sem_destroy(&accept_semaphore) == 0);
```
Comment (ADKaster) [inline Tests/Kernel/TestTCPSocket.cpp:59]: This is probably a good fix, but doing operations with side effects in a VERIFY() is no bueno. Even though the VERIFY is a release assertion, and will never be ifdef'd out, it still hurts readability and consistency.

```
rc = sem_wait(&accept_semaphore);
VERIFY(rc == 0);
```
is preferred for these (and in the thread func).

Makes we wonder if we have that pthread_barrier_wait thing.

Also I wonder if making the semaphore static is more idiomatic? Not super used to using posix thread primitives.

Comment (DanShaders) [reply Tests/Kernel/TestTCPSocket.cpp]: > but doing operations with side effects in a VERIFY() is no bueno

Fixed.

> Makes we wonder if we have that pthread_barrier_wait thing.

```shell
$ grep "pthread_barrier" -R Userland | wc -l
0
```

> Also I wonder if making the semaphore a file-scope static is more idiomatic?

Surely it isn't that important for a test file.

Comment (ADKaster) [reply Tests/Kernel/TestTCPSocket.cpp]: > Surely it isn't that important for a test file.

No, not really. but, if you don't take pride in your tests they'll become hard to update :)

**Review by ADKaster — APPROVED** [review]

**Comment by ADKaster** [comment]
Thanks for fixing this flaky test!

**Pull request merged**